feature: Implement gateway ratelimit

SubZero0 opened the pull request:
## Summary

This PR aims to add a ratelimiter to the gateway that currently has none except a `Task.Delay(5000)`.
It follows what's described in the [discord api docs](https://discord.com/developers/docs/topics/gateway#rate-limiting).

Fixes #1442 

## Changes

- Add a new type of bucket, `GatewayBucket`
- Add `IdentifyMaxConcurrency` property to `DiscordSocketConfig`
- Add `GetBotGatewayAsync` to `IDiscordClient`
- Add `session_start_limit` to `GetBotGatewayResponse`
- Sharded client will automatically deal with concurrent identify requests when possible

SubZero0 committed:
Implement gateway ratelimit

SubZero0 committed:
Remove unused code

SubZero0 committed:
Share WebSocketRequestQueue between clients

SubZero0 committed:
Add global limit and a way to change gateway limits

SubZero0 committed:
Refactoring variable to fit lib standards

Still34 commented on src/Discord.Net.Rest/Entities/Gateway/GatewayLimit.cs:
```diff
@@ -0,0 +1,23 @@
+    public struct GatewayLimit
+    {
+        /// <summary>
+        ///     The maximum amount of this type of request in a time window, that is set by <see cref="Seconds"/>.
```
Start with `Gets or sets` as the verb.

Still34 commented on src/Discord.Net.Rest/Entities/Gateway/GatewayLimit.cs:
```diff
@@ -0,0 +1,23 @@
+        /// </summary>
+        public int Count { get; set; }
+        /// <summary>
+        ///     The amount of seconds until the rate limiter resets the remaining requests <see cref="Count"/>.
```
Ditto.

Still34 commented on src/Discord.Net.Rest/Entities/Gateway/GatewayLimits.cs:
```diff
@@ -0,0 +1,29 @@
+        ///     Gets or sets the global limits for the gateway rate limiter.
+        /// </summary>
+        /// <remarks>
+        ///     It includes all the other limits, like Identify.
```
Use `This property` instead of `it`.

Still34 commented on src/Discord.Net.Rest/Net/Queue/GatewayBucket.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace Discord.Net.Queue
+{
+    public enum GatewayBucketType
```
Should this be in its own class?

SubZero0 replied:
I followed the same design as `ClientBucket` that has `ClientBucketType` in the same file, should that be changed?

Still34 replied:
Requesting input from other contributors.
cc @foxbot @Joe4evr

Joe4evr replied:
Idk about this, so I defer to fox's judgement.

Still34 commented on src/Discord.Net.WebSocket/DiscordSocketConfig.cs:
```diff
@@ -124,6 +125,16 @@ public class DiscordSocketConfig : DiscordRestConfig
 
+        /// <summary>
+        ///     Gets or sets the gateway limits.
+        ///     <note type="warning">
```
Move this to `<remarks>`. Replace `It` with `This property`.

SubZero0 replied:
I believe this warning should stay in the summary because VS doesn't show remarks (or it's disabled by default) and wanted to make who is using the lib aware that they shouldn't change the limits unless they know what they are doing, it might have really bad consequences.

Joe4evr replied:
VS 16.6 released just the other day which shows the Remarks section in Quick Info's. I didn't need to change any settings for that to come up.

SubZero0 replied:
Updated from 16.5 and it seems they added remarks, returns, parameters, etc to 16.6. Going to move it to remarks.

Still34 commented on src/Discord.Net.Rest/Entities/Gateway/GatewayLimits.cs:
```diff
@@ -0,0 +1,29 @@
+        /// </summary>
+        public GatewayLimit Identify { get; set; }
+
+        public GatewayLimits()
```
Ctor should include a summary of `Initializes a new...`

SubZero0 committed:
Update xml docs

Still34 commented on src/Discord.Net.Rest/Entities/Gateway/GatewayLimit.cs:
```diff
@@ -0,0 +1,23 @@
+        /// <summary>
+        ///     Gets or sets the maximum amount of this type of request in a time window, that is set by <see cref="Seconds"/>.
+        /// </summary>
+        public int Count { get; set; }
```
Might want to be clear with what the `int` here is with `<returns>`.

SubZero0 committed:
Update xml docs

SubZero0 committed:
Move warning to remarks

SubZero0 commented:
I'll need to do a few more changes to this PR

SubZero0 committed:
Remove specific RequestQueue for WebSocket and other changes

The only account limit is for identify that is dealt in a different way (exclusive semaphore), so websocket queues can be shared with REST and don't need to be shared between clients anymore.

Also added the ratelimit for presence updates.

SubZero0 commented:
After a bit more digging, I learned that the global queue is per-session, only Identify is per-account, and there's a "soft" presence update limit, so I changed my approach taking this into consideration.

- Added `IdentifySemaphoreName` and `PresenceUpdate` to `GatewayLimits` 
- Gateway will use the same `RequestQueue` as REST per client, won't be sharing one anymore
- Identify will use a named semaphore (the name is changeable at `GatewayLimits`), this will make possible to share the same semaphore between clients with or without a sharded client

I also noticed (if I understood it correctly) that if the user hits the gateway ratelimit, it won't be able to send heartbeats and will end losing the gateway connection either way. It would be possible to create a special limit for it but do we want to interfere this much?

SubZero0 committed:
Add summary to IdentifySemaphoreName

SubZero0 committed:
Fix spacing

SubZero0 commented:
While browsing the discord api docs github, I noticed this pull request: [https://github.com/discord/discord-api-docs/pull/1621](https://github.com/discord/discord-api-docs/pull/1621)

It seems discord provides a key that allows multiple shards to concurrently identify in [Get Gateway Bot](https://discord.com/developers/docs/topics/gateway#get-gateway-bot).

I'll see if there's a better way to implement it.

SubZero0 committed:
Add max_concurrency and other fixes

- Add session_start_limit to GetBotGatewayResponse
- Add GetBotGatewayAsync to IDiscordClient
- Add master/slave semaphores to enable concurrency
- Not store semaphore name as static
- Clone GatewayLimits when cloning the Config

SubZero0 committed:
Add missing RequestQueue parameter and wrong nullable

SubZero0 committed:
Add RequeueQueue paramater to Webhook

SubZero0 committed:
Better xml documentation

SubZero0 committed:
Remove GatewayLimits class and other changes

- Remove GatewayLimits
- Transfer a few properties to DiscordSocketConfig
- Remove unnecessary usings

SubZero0 committed:
Remove unnecessary using and wording

SubZero0 committed:
Remove more unnecessary usings

SubZero0 commented:
After some talk and more consideration, I reached the conclusion that Discord wouldn't change specific limits for the gateway since what needs to be changed is sent by REST (max concurrency) or can be solved by adding more shards (gateway global limits).

Taking this into consideration, I removed `GatewayLimits`.
I left the semaphore-related properties in `DiscordSocketConfig`, I'm also considering removing these since the limited use-case would be automatic identify control by independent `DiscordSocketClient`s (without a `DiscordShardedClient` as mediator).

SubZero0 committed:
Change named Semaphores to SemaphoreSlim

SubZero0 commented:
I decided to remove the named semaphore since it would add a lot more complexity. If someone is sharding for themselves, they have entire control when a client is going to identify.

Now only `IdentifyMaxConcurrency` is present at the config, since it's needed to create the "slave semaphores" for concurrent identify when automatic sharding is disabled.

SubZero0 committed:
Remove unused using

SubZero0 committed:
Update branch

SubZero0 committed:
Merge branch 'dev' into gateway-ratelimit

SubZero0 committed:
Fix merge conflicts and update to new ratelimit

SubZero0 committed:
Merge branch 'dev' into gateway-ratelimit

SubZero0 committed:
Fixing merge, ignore limit for heartbeat, and dispose

SubZero0 committed:
Missed one place and better xml docs.

SubZero0 committed:
Wait identify before opening the connection

SubZero0 committed:
Only request identify ticket when needed

SubZero0 committed:
Move identify control to sharded client

SubZero0 committed:
Better description for IdentifyMaxConcurrency

SubZero0 committed:
Add lock to InvalidSession
